src/storage/src/source/reclock.rs

```diff
@@ -340,50 +339,56 @@ where
     }
 
     pub fn compact(&mut self, new_since: Antichain<IntoTime>) {
-        let mut inner = &mut *self.inner.borrow_mut();
         // Ignore compaction requests while we initialize
-        if !PartialOrder::less_than(&inner.as_of, &inner.upper) {
+        if !self.initialized() {
             return;
         }
-        if !PartialOrder::less_equal(&inner.since, &new_since) {
+        let inner = &mut *self.inner.borrow_mut();
+        if !PartialOrder::less_equal(&self.since, &new_since) {
             panic!(
                 "ReclockFollower: `new_since` ({:?}) is not beyond \
                 `self.since` ({:?}).",
-                new_since, inner.since,
+                new_since, self.since,
             );
         }
+        inner.since.update_iter(
+            self.since
+                .iter()
+                .map(|t| (t.clone(), -1))
+                .chain(new_since.iter().map(|t| (t.clone(), 1))),
+        );
+        self.since = new_since;
+
         // Compact the remap trace according to the computed frontier
         for (_src_ts, ts, _diff) in inner.remap_trace.iter_mut() {
-            ts.advance_by(new_since.borrow());
+            ts.advance_by(inner.since.frontier());
         }
         // And then consolidate
         consolidation::consolidate_updates(&mut inner.remap_trace);
-        assert_eq!(new_since.len(), 1);
-        inner.source_since = Antichain::from_iter(
-            new_since
-                .get(0)
-                .map(|since_ts| {
-                    inner.remap_trace.iter().filter_map(|(src_ts, ts, _diff)| {
-                        if PartialOrder::less_equal(ts, since_ts) {
-                            Some(src_ts.clone())
-                        } else {
-                            None
-                        }
-                    })
-                })
-                .into_iter()
-                .flatten(),
-        );
-        inner.since = new_since;
     }
 
     pub fn share(&self) -> Self {
+        self.inner
```

Comment:
I have seen this pattern in different places now, funny that we don't have a re-usable impl somewhere yet...